File: CousinBinaryTree.py
#993. Cousins in Binary Tree
"""
Time Complexity : O(n)
Space Complexity : O(n)
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        
        q = deque()
        q.append(root)
        result = []
        
        while q :
            size = len(q)
            #print(size)
            hs = set() #if x and y are in hs then they are at saem level
            for i in range(0, size) :
                n = q.popleft()
                
                if n.left != None and n.right != None :
                    if n.left.val in [x, y] and n.right.val in [x, y]:
                        return False
                
                if n.left != None :
                    hs.add(n.left.val)
                    q.append(n.left)
                
                if n.right != None :
                    hs.add(n.right.val)
                    q.append(n.right)       
            
            if x in hs and y in hs:
                return True
                    
        return False

File: CousinBinaryTree.cpp
//Time Complexity:O(n)
//Space Complexity:O(n)
//Performed BFS to keep track of the node as well the parent node and check if x and y are on the same level and if their parents are
//different, if yes, return true, else return false
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) { 
        queue<TreeNode*> q;
        queue<TreeNode*> pq;

        q.push(root);
        pq.push(NULL);

        bool x_found = false;
        bool y_found = false;
        TreeNode* xParent = NULL;
        TreeNode* yParent = NULL;

        while(!q.empty()){
            int size = q.size();
             
            for(int i=0;i<size;i++){
                TreeNode* curr = q.front();
                TreeNode* pCurr = pq.front();
                
                q.pop();
                pq.pop();
               

                if(curr->left!=NULL){
                    q.push(curr->left);
                    pq.push(curr);

                }
                if(curr->right!=NULL){
                    q.push(curr->right);
                    pq.push(curr);
                }

                if(curr->val == x){
                    x_found = true;
                    xParent = pCurr;
                }
                if(curr->val == y){
                    y_found = true;
                    yParent = pCurr;


                }

                




            }

            if((x_found && !y_found)||(y_found && !x_found)){
                return false;
            }
            
            if((x_found && y_found)&&(xParent != yParent)){
                return true;
                
            }


            
            


        }

        return false;

        
    }
};